samples/sample_multi_transcode/src/sample_multi_transcode.cpp

```diff
@@ -149,146 +150,194 @@ mfxStatus Launcher::Init(int argc, msdk_char *argv[])
     {
+        /* In the case of joined sessions, need to create device only for a zero session
+         * In the case of a shared buffer, need to create device only for decode */
+        if ((m_InputParamsArray[i].bIsJoin && i != 0) || m_InputParamsArray[i].eMode == Source)
+            bNeedToCreateDevice = false;
+
+#if defined(_WIN32) || defined(_WIN64)
+        ForceImplForSession(i);
 
-        m_pAllocParam.reset(new D3D11AllocatorParams);
-        m_hwdev.reset(new CD3D11Device());
-        /* The last param set in vector always describe VPP+ENCODE or Only VPP
-         * So, if we want to do rendering we need to do pass HWDev to CTranscodingPipeline */
-        if (m_InputParamsArray[m_InputParamsArray.size() -1].eModeExt == VppCompOnly)
+        if (m_eDevType == MFX_HANDLE_D3D9_DEVICE_MANAGER)
         {
-            /* Rendering case */
-            sts = m_hwdev->Init(NULL, 1,
-                MSDKAdapter::GetNumber(0,MFX_IMPL_VIA_D3D11 | MFX_IMPL_BASETYPE(m_InputParamsArray[0].libType)) );
-            m_InputParamsArray[m_InputParamsArray.size() -1].m_hwdev = m_hwdev.get();
+            if (bNeedToCreateDevice)
+            {
+                mfxAllocatorParams* pAllocParam(new D3DAllocatorParams());
+                D3DAllocatorParams* pD3DParams = dynamic_cast<D3DAllocatorParams*>(pAllocParam);
+                std::unique_ptr<CHWDevice> hwdev(new CD3D9Device());
+
+                /* The last param set in vector always describe VPP+ENCODE or Only VPP
+                 * So, if we want to do rendering we need to do pass HWDev to CTranscodingPipeline */
+                if (m_InputParamsArray[m_InputParamsArray.size() - 1].eModeExt == VppCompOnly)
+                {
+                    /* Rendering case */
+                    sts = hwdev->Init(NULL, 1,
+                        MSDKAdapter::GetNumber(0, MFX_IMPL_VIA_D3D9 | MFX_IMPL_BASETYPE(m_InputParamsArray[i].libType)));
+                    m_InputParamsArray[m_InputParamsArray.size() - 1].m_hwdev = hwdev.get();
+                }
+                else /* NO RENDERING */
+                {
+                    sts = hwdev->Init(NULL, 0,
+                        MSDKAdapter::GetNumber(0, MFX_IMPL_VIA_D3D9 | MFX_IMPL_BASETYPE(m_InputParamsArray[i].libType)));
+                }
+                MSDK_CHECK_STATUS(sts, "hwdev->Init failed");
+                sts = hwdev->GetHandle(MFX_HANDLE_D3D9_DEVICE_MANAGER, (mfxHDL*)&hdl);
+                MSDK_CHECK_STATUS(sts, "hwdev->GetHandle failed");
+                // set Device Manager to external dx9 allocator
+                pD3DParams->pManager = (IDirect3DDeviceManager9*)hdl;
+
+                m_pAllocParams.push_back(std::shared_ptr<mfxAllocatorParams>(pAllocParam));
+                m_hwdevs.push_back(std::move(hwdev));
+                hdls.push_back(hdl);
+            }
+            else
+            {
+                m_pAllocParams.push_back(m_pAllocParams.back());
+                hdls.push_back(hdls.back());
```

Comment:
Not sure that I've get the idea here. We creating a copy of shared pointer, thus pinning object. Btw, are we sure, that we have at least one element in both m_pAllocParams and hdls?

Reply:
> Not sure that I've get the idea here. We creating a copy of shared pointer, thus pinning object.

Yes, that's right. For example, in the case of 1-N, the object for alloc parameters will be created once for the zero session. For the remaining sessions, this object will be reused. Instead of creating new objects, the number of references to the first one will increase.

> Btw, are we sure, that we have at least one element in both m_pAllocParams and hdls?

I added checks